### PR TITLE
Reorganized layout of project

### PR DESCRIPTION
**Pull request by codeinred**
Split project into header files and example programs

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,17 +1,83 @@
...
-    src/stream.hpp
-    src/stream.cpp
-    src/task.hpp
-    src/task.cpp
-    src/eff.hpp
-    src/eff.cpp)
+# We want to export compile commands b/c it's useful for ccls
+set(CMAKE_EXPORT_COMPILE_COMMANDS true)
 
-set_property(TARGET cxxeffect PROPERTY CXX_STANDARD 17)
```
Comment (s5bug) [inline CMakeLists.txt:14]: Why was this removed? Isn't it helpful for IDEs and such?

File: examples/cat.cpp [hunk 2]
```diff
@@ -1,6 +1,6 @@
-#include "eff.hpp"
-#include "stream.hpp"
-#include "task.hpp"
+// #include <cxxeffect/eff.hpp>
```
Comment (s5bug) [inline examples/cat.cpp]: If it's not needed, it should just be removed. Alternatively, perhaps `eff.hpp` should import both `stream` and `task` for you?

File: CMakeLists.txt (hunk 1 above)
Comment (codeinred) [reply CMakeLists.txt:14]: We could add it back in, although it'd have to be targeted towards C++20 since we'll be integrating coroutines!

Comment (codeinred) [reply CMakeLists.txt:14]: CCLS reads the build configuration from compile_commands.json so it doesn't need set_property, although if you know of other IDEs that rely on it, retaining it could be a good idea

File: examples/cat.cpp (hunk 2 above)
Comment (codeinred) [reply examples/cat.cpp]: This was an oversight; I'll remove it

File: CMakeLists.txt (hunk 1 above)
Comment (s5bug) [reply CMakeLists.txt:14]: VSCode's CMake stuff relies on it by default. CLion also relies on it 😄

File: examples/cat.cpp (hunk 2 above)
Comment (s5bug) [reply examples/cat.cpp]: What do you think about `eff.hpp` including all of the library headers by default? That would make SFH generation easier too, wouldn't it?

Comment (codeinred) [reply examples/cat.cpp]: > If it's not needed, it should just be removed. Alternatively, perhaps `eff.hpp` should import both `stream` and `task` for you?

I removed this include!

File: CMakeLists.txt (hunk 1 above)
Comment (s5bug) [inline CMakeLists.txt:14]: Alright, so just adding back the property for CXX20 should be good here. I'm not a CMake expert, would that negate the need for the `-std`/`/std` compiler flags? I feel like it should.

**Review by s5bug — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 1 above)
Comment (codeinred) [reply CMakeLists.txt:14]: Probably! I'll test it!

Comment (codeinred) [reply CMakeLists.txt:14]: Update: We still need the -std / std compile flags. CMake doesn't add `-std=c++20` even when we have `set_property(TARGET cxxeffect PROPERTY CXX_STANDARD 20)`

Comment (s5bug) [reply CMakeLists.txt:14]: This seems relevant: https://stackoverflow.com/a/20165519/4580217

Esp. for `-fcoroutines`

**Comment by codeinred** [comment]
I added back set_property per your request!

File: CMakeLists.txt (hunk 1 above)
Comment (s5bug) [reply CMakeLists.txt:14]: CMake doesn't have a feature flag for coroutines, and `target_compile_features(cxxeffect PRIVATE cxx_std_20)` should be automatically added when `CXX_STANDARD` is set to 20...

Comment (s5bug) [reply CMakeLists.txt:14]: Ah, what happens when you set `CXX_STANDARD_REQUIRED` instead of `CXX_STANDARD`?

Comment (codeinred) [reply CMakeLists.txt:14]: That still doesn't do it, but if we require C++20 as part of the interface of cxxeffect it works! i will update CMakeLists.txt.
```cmake
target_compile_features(cxxeffect INTERFACE cxx_std_20)
```

**Comment by s5bug** [comment]
Unrelated to the PR content, but to prevent "Merged x into x" commits, you should be `rebase`ing instead of `pull`ing (you can do both at once with `git pull --rebase` on the command line; no clue how to do it in GitKraken).

File: CMakeLists.txt (hunk 1 above)
Comment (s5bug) [reply CMakeLists.txt:14]: Alright, we should do both `CXX_STANDARD_REQUIRED` and `cxx_std_20` then 👍 

Really odd that this happens... I feel like CMake should be doing the latter from the former.

**Comment by codeinred** [comment]
Oh! I appreciate the tip about `git pull --rebase` - I didn't know about it!

File: CMakeLists.txt [hunk 3]
```diff
@@ -1,17 +1,84 @@
...
+# cxxeffect is an interface library b/c it has a lot of templated code
+# That means it's best implemented as header-only without the need for cpp files
+add_library(cxxeffect INTERFACE)
 
-target_link_libraries(cxxeffect pthread)
+set_property(TARGET cxxeffect PROPERTY CXX_STANDARD_REQUIRED 20)
 
+target_compile_features(cxxeffect INTERFACE cxx_std_20)
+
+# boilerplate compiler options to have support for C++20. We can remove if you don't want them.
```
Comment (s5bug) [inline CMakeLists.txt]: Comment no longer needed here, I think?

**Comment by codeinred** [comment]
Added:
```cmake
set_property(TARGET cxxeffect PROPERTY CXX_STANDARD_REQUIRED 20)

target_compile_features(cxxeffect INTERFACE cxx_std_20)
```
And removed the target_compile_options that were made redundant.

If there are no other issues I think we're good to merge!

File: CMakeLists.txt [hunk 4]
```diff
@@ -1,17 +1,84 @@
...
+if (MSVC)
+    target_compile_options(cxxeffect INTERFACE "/EHsc")
```
Comment (s5bug) [inline CMakeLists.txt]: What does `/EHsc` do? I'd add a comment here to explain to people like me who aren't familiar with MSVC.

File: CMakeLists.txt [hunk 5]
```diff
@@ -1,17 +1,84 @@
...
+    cxxeffect
+    INTERFACE
+    $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include/>
+    $<INSTALL_INTERFACE:include>
+)
+
+# Provides install target, so users can install the library on their system
+# if they like
+
+option(CXXEFFECT_Install "Install CMake targets during install step." ON)
```
Comment (s5bug) [inline CMakeLists.txt]: Doesn't `cmake -B` have an option to choose the build target? If this is a separate build step, make it a build target if CMake can do that (`--target install`).

Or, is it a flag for i.e. locating a CMake package? Should it then be something like `CXXEFFECT_INSTALL_CMAKE_FILES`? (I have to believe there's some community standard for it, but it seems odd to me to not be in all-caps)

File: CMakeLists.txt [hunk 6]
```diff
@@ -1,17 +1,84 @@
...
+# Use pthreads, if availible
+set(THREADS_PREFER_PTHREAD_FLAG ON)
+# Enable threads in project
+find_package(Threads REQUIRED)
+
+# If we're building this project, and not something that has this project as a dependency
+# Then we want to build files in the examples directory (which contains the cat)
+# This will compile every file in the examples directory as a separate program
+if(CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME)
+    file(GLOB all_tests "examples/*.cpp")
```
Comment (s5bug) [inline CMakeLists.txt]: Rename to `all_examples`

File: CMakeLists.txt [hunk 7]
```diff
@@ -1,17 +1,84 @@
...
+
+# Use pthreads, if availible
+set(THREADS_PREFER_PTHREAD_FLAG ON)
+# Enable threads in project
+find_package(Threads REQUIRED)
+
+# If we're building this project, and not something that has this project as a dependency
+# Then we want to build files in the examples directory (which contains the cat)
+# This will compile every file in the examples directory as a separate program
+if(CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME)
```
Comment (s5bug) [inline CMakeLists.txt:53]: What do you think about a flag `CXXEFFECT_BUILD_EXAMPLES` that's by-default `ON`, and gets &&ed with this?

**Review by s5bug — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 4 above)
Comment (codeinred) [reply CMakeLists.txt]: It makes msvc exception handling standards-compliant. Otherwise, msvc won't do stack unwinding

Comment (codeinred) [reply CMakeLists.txt]: Reviewing https://docs.microsoft.com/en-us/cpp/build/reference/eh-exception-handling-model?view=msvc-160, we should perhaps change it to "/EHa"

File: CMakeLists.txt (hunk 3 above)
Comment (codeinred) [reply CMakeLists.txt]: Yes. I'll add comments explaining the additional compiler options like "/EH" and remove that comment.

File: CMakeLists.txt (hunk 5 above)
Comment (codeinred) [reply CMakeLists.txt]: I think some of the changes we're making are starting to fall outside of the scope of this pull request. I'm going to push the updated CMake documentation. How about we merge this pull request after that, and then create a separate branch for build system updates?

File: CMakeLists.txt (hunk 7 above)
Comment (codeinred) [reply CMakeLists.txt:53]: I've updated all_tests to all_examples, and added documentation for the "/EH" flag. I think some of these changes and improvements should be made on a separate branch specifically for the build system. Do you think that would work?

File: CMakeLists.txt (hunk 5 above)
Comment (s5bug) [reply CMakeLists.txt]: Agreed. Things should follow the naming conventions though, so I believe this should still be `CXXEFFECT_INSTALL`, but I could be entirely wrong.

Comment (codeinred) [reply CMakeLists.txt]: Huh. It turns out that that option is completely unnecessary for installation to work. Even with that line completely removed, you can still install via running
```bash
sudo cmake --install .
```
In the build directory

Comment (codeinred) [reply CMakeLists.txt]: I'm just gonna remove it

Comment (codeinred) [reply CMakeLists.txt]: Update: I removed it since it didn't do anything. Are we ready to merge into main?

**Review by s5bug — APPROVED** [review]
Thank you so much for your work on this!

**Comment by codeinred** [comment]
Thank you so much for being so patient with me!
